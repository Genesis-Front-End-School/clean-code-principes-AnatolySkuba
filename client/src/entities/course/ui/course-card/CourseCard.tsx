import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { VideoPlayer, Rating } from 'courses-components';

import { ICourse } from '../..';

function CourseCard({
  id,
  title,
  description,
  previewImageLink,
  lessonsCount,
  rating,
  meta,
}: ICourse): JSX.Element {
  const [isHovered, setIsHovered] = useState(false);

  return (
    <li
      className="relative min-w-fill rounded-lg bg-white dark:bg-stone-900 duration-700 border shadow-md hover:shadow-lg shadow-gray-500 hover:shadow-gray-800 dark:hover:shadow-gray-400"
      onMouseEnter={() => setIsHovered(true)}
      onMouseLeave={() => setIsHovered(false)}
    >
      <Link to={id} className="block p-3 pb-10">
        <div className="relative h-28 w-64 mx-auto">
          {isHovered ? (
            <VideoPlayer
              className="absolute top-0 h-full w-full rounded-md object-contain"
              src={meta.courseVideoPreview?.link}
              id={id}
              muted
              autoPlay
            />
          ) : (
            <img
              alt={title}
              src={`${previewImageLink}/cover.webp`}
              className="absolute top-0 h-full w-full rounded-md object-contain"
            />
          )}
        </div>
        <div className="mt-1 dark:text-gray-200">
          <h2 className="font-medium">{title}</h2>
          <p className="text-sm text-gray-500 dark:text-gray-200">{description}</p>

          <div className="mt-1 text-xs ">
            <div>
              {meta.skills && (
                <p className="text-gray-500 dark:text-gray-200">
                  {`Skills: ${meta.skills.join(', ')}.`}
                </p>
              )}
              <div className="absolute bottom-3 w-full flex justify-between">
                <Rating rating={rating} />
                <p className="mr-6 text-gray-500 dark:text-gray-400">
                  Lessons:
                  {lessonsCount}
                </p>
              </div>
            </div>
          </div>
        </div>
      </Link>
    </li>
  );
}

export default CourseCard;
