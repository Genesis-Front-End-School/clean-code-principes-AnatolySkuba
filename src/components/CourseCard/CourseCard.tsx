import React, { useEffect, useState } from "react";
import Hls from "hls.js";
import { FcAcceptDatabase, FcPositiveDynamic, FcVoicePresentation } from "react-icons/fc";

import { Course } from "../../utils/types";
import { ROUTER_KEYS, STORAGE_KEYS } from "../../utils/constants";

export const CourseCard = ({
    id,
    title,
    description,
    previewImageLink,
    lessonsCount,
    rating,
    meta,
}: Course): JSX.Element => {
    const [videoEl, setVideoEl] = useState<HTMLVideoElement | null>(null);
    const [hlsEl, setHlsEl] = useState<Hls | null>(null);
    const [isPlaying, setIsPlaying] = useState(false);

    useEffect(() => {
        const video = document.getElementById(id) as HTMLVideoElement;
        !videoEl && setVideoEl(video);
    }, [id, videoEl]);

    const handleOnMouseOver = () => {
        const savedTime = localStorage.getItem(`currentTime ${id}`);

        if (
            !savedTime &&
            Hls.isSupported() &&
            meta.courseVideoPreview?.link &&
            localStorage.getItem(STORAGE_KEYS.TOKEN)
        ) {
            const hls = new Hls({
                xhrSetup: (xhr) => {
                    xhr.responseType = "json";
                    xhr.setRequestHeader(
                        "Authorization",
                        `Bearer ${localStorage.getItem(STORAGE_KEYS.TOKEN)}`
                    );
                },
            });
            hls.loadSource(meta.courseVideoPreview?.link);
            // hls.loadSource("https://test-streams.mux.dev/x36xhzz/x36xhzz.m3u8");
            setHlsEl(hls);
            videoEl && hls.attachMedia(videoEl);
            !savedTime && videoEl && videoEl.load();
        } else {
            if (!hlsEl) {
                localStorage.removeItem(`currentTime ${id}`);
                handleOnMouseOver();
            }
            videoEl && hlsEl && hlsEl.startLoad(Number(savedTime));
        }

        if (videoEl && !isPlaying) {
            videoEl.play();
            setIsPlaying(true);
        }
    };

    const handleOnMouseOut = () => {
        if (videoEl && hlsEl && isPlaying) {
            videoEl?.currentTime &&
                localStorage.setItem(`currentTime ${id}`, videoEl?.currentTime.toString());
            videoEl.pause();
            hlsEl?.stopLoad();
            setIsPlaying(false);
        }
    };

    return (
        <li className="block min-w-80 rounded-lg p-3 border shadow-md shadow-gray-500 hover:scale-105">
            <a href={`${ROUTER_KEYS.COURSE}/${id}`}>
                <div className="relative h-28 w-64 mx-auto group" data-testid="group">
                    <img
                        alt={title}
                        src={`${previewImageLink}/cover.webp`}
                        className="absolute top-0 h-full w-full rounded-md object-cover group-hover:hidden"
                    />
                    <video
                        id={id}
                        className="absolute top-0 h-full w-full opacity-0 rounded-md object-cover group-hover:opacity-100"
                        loop
                        preload="none"
                        muted
                        onMouseOver={handleOnMouseOver}
                        onMouseOut={handleOnMouseOut}
                        data-testid="video-player"
                    />
                </div>
                <div className="mt-1">
                    <p className="font-medium">{title}</p>
                    <p className="text-sm text-gray-500">{description}</p>

                    <div className="flex items-center gap-3 mt-1 text-xs">
                        <div>
                            <p className="text-gray-500">Lessons</p>
                            <p className="flex font-medium gap-1">
                                <FcAcceptDatabase size="16" />
                                {lessonsCount}
                            </p>
                        </div>

                        <div>
                            <p className="text-gray-500">Rating</p>
                            <p className="flex font-medium gap-1">
                                <FcPositiveDynamic size="16" />
                                {rating}
                            </p>
                        </div>

                        <div>
                            <div className="flex gap-2">
                                <p className="text-gray-500">Skills</p>
                                <FcVoicePresentation size="16" />
                            </div>
                            <ul>
                                {meta.skills ? (
                                    meta.skills?.map((skill, index) => (
                                        <li key={index}>
                                            <p className="w-50 font-medium">{skill}</p>
                                        </li>
                                    ))
                                ) : (
                                    <li>
                                        <p>?</p>
                                    </li>
                                )}
                            </ul>
                        </div>
                    </div>
                </div>
            </a>
        </li>
    );
};
