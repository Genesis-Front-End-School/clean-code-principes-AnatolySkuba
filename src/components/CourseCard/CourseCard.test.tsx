import { render, screen } from "@testing-library/react";
import { CourseCard } from "./CourseCard";

describe("CourseCard component", () => {
    const createComponent = () =>
        render(
            <CourseCard
                id={""}
                title={""}
                description={""}
                previewImageLink={""}
                lessonsCount={0}
                rating={0}
                meta={{
                    skills: [],
                    courseVideoPreview: {
                        link: "",
                    },
                }}
            />
        );

    beforeEach(createComponent);

    test("Render CourseCard correctly", () => {
        expect(screen.getByRole("list")).toBeInTheDocument();
        expect(screen.getByRole("listitem")).toBeInTheDocument();
        expect(screen.getByRole("link")).toBeInTheDocument();
        expect(screen.getByRole("img")).toBeInTheDocument();
        expect(screen.getByTestId("video-player")).toBeInTheDocument();
        expect(screen.getByText("Lessons")).toBeInTheDocument();
        expect(screen.getByText("Rating")).toBeInTheDocument();
        expect(screen.getByText("Skills")).toBeInTheDocument();
    });

    test("List snapshot", () => {
        expect(screen.getByRole("listitem")).toMatchSnapshot();
    });
});
