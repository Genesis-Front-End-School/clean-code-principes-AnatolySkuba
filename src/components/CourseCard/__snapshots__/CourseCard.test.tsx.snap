// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCard component List snapshot 1`] = `
<li
  class="block min-w-80 rounded-lg p-3 border shadow-md shadow-gray-500 hover:scale-105"
>
  <a
    href="course/"
  >
    <div
      class="relative h-28 w-64 mx-auto group"
      data-testid="group"
    >
      <img
        alt=""
        class="absolute top-0 h-full w-full rounded-md object-cover group-hover:hidden"
        src="/cover.webp"
      />
      <video
        class="absolute top-0 h-full w-full opacity-0 rounded-md object-cover group-hover:opacity-100"
        data-testid="video-player"
        id=""
        loop=""
        preload="none"
      />
    </div>
    <div
      class="mt-1"
    >
      <p
        class="font-medium"
      />
      <p
        class="text-sm text-gray-500"
      />
      <div
        class="flex items-center gap-3 mt-1 text-xs"
      >
        <div>
          <p
            class="text-gray-500"
          >
            Lessons
          </p>
          <p
            class="flex font-medium gap-1"
          >
            <svg
              enable-background="new 0 0 48 48"
              fill="currentColor"
              height="16"
              stroke="currentColor"
              stroke-width="0"
              version="1"
              viewBox="0 0 48 48"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="#D1C4E9"
              >
                <path
                  d="M38,7H10C8.9,7,8,7.9,8,9v6c0,1.1,0.9,2,2,2h28c1.1,0,2-0.9,2-2V9C40,7.9,39.1,7,38,7z"
                />
                <path
                  d="M38,19H10c-1.1,0-2,0.9-2,2v6c0,1.1,0.9,2,2,2h28c1.1,0,2-0.9,2-2v-6C40,19.9,39.1,19,38,19z"
                />
                <path
                  d="M38,31H10c-1.1,0-2,0.9-2,2v6c0,1.1,0.9,2,2,2h28c1.1,0,2-0.9,2-2v-6C40,31.9,39.1,31,38,31z"
                />
              </g>
              <circle
                cx="38"
                cy="38"
                fill="#43A047"
                r="10"
              />
              <polygon
                fill="#DCEDC8"
                points="42.5,33.3 36.8,39 34.1,36.3 32,38.5 36.8,43.3 44.6,35.5"
              />
            </svg>
            0
          </p>
        </div>
        <div>
          <p
            class="text-gray-500"
          >
            Rating
          </p>
          <p
            class="flex font-medium gap-1"
          >
            <svg
              enable-background="new 0 0 48 48"
              fill="currentColor"
              height="16"
              stroke="currentColor"
              stroke-width="0"
              version="1"
              viewBox="0 0 48 48"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="#00BCD4"
              >
                <rect
                  height="20"
                  width="10"
                  x="19"
                  y="22"
                />
                <rect
                  height="34"
                  width="10"
                  x="32"
                  y="8"
                />
                <rect
                  height="12"
                  width="10"
                  x="6"
                  y="30"
                />
              </g>
              <g
                fill="#3F51B5"
              >
                <polygon
                  points="11,8 21,18 21,8"
                />
                <rect
                  height="14.1"
                  transform="matrix(-.707 -.707 .707 -.707 10.879 36.506)"
                  width="4"
                  x="11"
                  y="8.9"
                />
              </g>
            </svg>
            0
          </p>
        </div>
        <div>
          <div
            class="flex gap-2"
          >
            <p
              class="text-gray-500"
            >
              Skills
            </p>
            <svg
              enable-background="new 0 0 48 48"
              fill="currentColor"
              height="16"
              stroke="currentColor"
              stroke-width="0"
              version="1"
              viewBox="0 0 48 48"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M40,22h-8l-4,4V12c0-2.2,1.8-4,4-4h8c2.2,0,4,1.8,4,4v6C44,20.2,42.2,22,40,22z"
                fill="#2196F3"
              />
              <circle
                cx="17"
                cy="19"
                fill="#FFA726"
                r="8"
              />
              <path
                d="M30,36.7c0,0-3.6-6.7-13-6.7S4,36.7,4,36.7V40h26V36.7z"
                fill="#607D8B"
              />
            </svg>
          </div>
          <ul />
        </div>
      </div>
    </div>
  </a>
</li>
`;
